File: ecs-hosted/go/application/service/migrationsService.go
package service

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/pulumi/pulumi-aws/sdk/v7/go/aws/cloudwatch"
	"github.com/pulumi/pulumi-aws/sdk/v7/go/aws/ec2"
	"github.com/pulumi/pulumi-aws/sdk/v7/go/aws/ecs"
	"github.com/pulumi/pulumi-aws/sdk/v7/go/aws/iam"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/infrastructure/application/config"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/infrastructure/application/utils"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/infrastructure/common"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

const cpu = 256
const memoryReservation = 512

func NewMigrationsService(ctx *pulumi.Context, name string, args *MigrationsContainerServiceArgs, opts ...pulumi.ResourceOption) (*MigrationsContainerService, error) {
	var resource MigrationsContainerService

	// create our parented options
	options := append(opts, pulumi.Parent(&resource))

	err := ctx.RegisterComponentResource("pulumi:dbMigrations", name, &resource, opts...)
	if err != nil {
		return nil, err
	}

	role, err := NewEcsRole(ctx, fmt.Sprintf("%s-role", name), args.Region, nil, options...)
	if err != nil {
		return nil, err
	}

	policyArn := common.GetIamPolicyArn(args.Region, string(iam.ManagedPolicyAmazonECSTaskExecutionRolePolicy))

	_, err = iam.NewRolePolicyAttachment(ctx, fmt.Sprintf("%s-task-role-pol", name), &iam.RolePolicyAttachmentArgs{
		Role:      role,
		PolicyArn: pulumi.String(policyArn),
	}, options...)

	if err != nil {
		return nil, err
	}

	doc, err := NewSecretsManagerPolicy(ctx, name, args.Region, args.SecretsManagerPrefix, args.KmsServiceKeyId, args.AccountId, options...)
	if err != nil {
		return nil, err
	}

	_, err = iam.NewRolePolicy(ctx, fmt.Sprintf("%s-secret-pol", name), &iam.RolePolicyArgs{
		Role:   role,
		Policy: doc,
	})

	if err != nil {
		return nil, err
	}

	egress := args.SecurityGroupEgressRules

	if args.EnablePrivateLoadBalancerAndLimitEgress {
		egress = append(egress, ec2.SecurityGroupEgressArgs{
			FromPort:       pulumi.Int(443),
			ToPort:         pulumi.Int(443),
			Protocol:       pulumi.String("TCP"),
			SecurityGroups: pulumi.StringArray{args.VpcEndpointSecurityGroupId},
			Description:    pulumi.String("Allow egress from ecs service to VPC Endpoint"),
		}, ec2.SecurityGroupEgressArgs{
			FromPort:      pulumi.Int(443),
			ToPort:        pulumi.Int(443),
			Protocol:      pulumi.String("TCP"),
			PrefixListIds: pulumi.StringArray{args.PrefixListId},
			Description:   pulumi.String("Allow egress from ecs service to S3 VPC Endpoint"),
		})
	} else {
		egress = append(egress, ec2.SecurityGroupEgressArgs{
			FromPort:    pulumi.Int(0),
			ToPort:      pulumi.Int(0),
			Protocol:    pulumi.String("-1"),
			CidrBlocks:  pulumi.ToStringArray([]string{"0.0.0.0/0"}),
			Description: pulumi.String("Allows egress to all IP addresses"),
		})
	}

	resource.SecurityGroup, err = ec2.NewSecurityGroup(ctx, fmt.Sprintf("%s-sg", name), &ec2.SecurityGroupArgs{
		VpcId:  args.VpcId,
		Egress: egress,
	}, options...)

	if err != nil {
		return nil, err
	}

	cluster, err := ecs.NewCluster(ctx, fmt.Sprintf("%s-cluster", name), &ecs.ClusterArgs{}, options...)
	if err != nil {
		return nil, err
	}

	ecrAccountId := args.AccountId
	if args.EcrRepoAccountId != "" {
		ecrAccountId = args.EcrRepoAccountId
	}

	imageName := fmt.Sprintf("pulumi/migrations:%s", args.ImageTag)
	fullQualifiedImage := utils.NewEcrImageTag(ecrAccountId, args.Region, imageName, args.ImagePrefix)

	containerDef, err := newContainerDefinitions(ctx, "migrations-task", args, fullQualifiedImage, options...)
	if err != nil {
		return nil, err
	}

	taskDefinition, err := ecs.NewTaskDefinition(ctx, fmt.Sprintf("%s-task-def", name), &ecs.TaskDefinitionArgs{
		Family:                  pulumi.String("pulumi-migration-task"),
		NetworkMode:             pulumi.String("awsvpc"),
		Cpu:                     pulumi.String(fmt.Sprintf("%d", cpu)),
		Memory:                  pulumi.String(fmt.Sprintf("%d", memoryReservation)),
		RequiresCompatibilities: pulumi.StringArray{pulumi.String("FARGATE")},
		ExecutionRoleArn:        role.Arn,
		ContainerDefinitions:    containerDef,
	}, options...)

	if err != nil {
		return nil, err
	}

	sgOptions := append(options, pulumi.DeleteBeforeReplace(true))
	_, err = ec2.NewSecurityGroupRule(ctx, fmt.Sprintf("%s-migrations-to-db-rule", name), &ec2.SecurityGroupRuleArgs{
		Type:                  pulumi.String("ingress"),
		SecurityGroupId:       args.DatabaseArgs.SecurityGroupId,
		SourceSecurityGroupId: resource.SecurityGroup.ID(),
		FromPort:              pulumi.Int(3306),
		ToPort:                pulumi.Int(3306),
		Protocol:              pulumi.String("TCP"),
	}, sgOptions...)

	if err != nil {
		return nil, err
	}

	if ctx.DryRun() {
		ctx.Log.Info("Skipping database migration task on Pulumi Preview", nil)
		return nil, nil
	}

	if !args.ExecuteMigrations {
		ctx.Log.Info("Skipping database migration based on PULUMI_EXECUTE_MIGRATIONS env var set to 'false'", nil)
		return nil, nil
	}

	_ = pulumi.All(
		cluster.ID(),
		resource.SecurityGroup.ID(),
		args.PrivateSubnetIds,
		taskDefinition.Arn,
		taskDefinition.Family,
	).ApplyT(func(applyArgs []any) string {
		clusterId := applyArgs[0].(pulumi.ID)
		sgId := applyArgs[1].(pulumi.ID)
		privateSubnets := applyArgs[2].([]string)
		taskDefArn := applyArgs[3].(string)
		taskDefFamily := applyArgs[4].(string)

		err := NewDatabaseMigrationTask(ctx, &MigrationTaskArgs{
			ContainerBaseArgs: &args.ContainerBaseArgs,
			Cluster:           string(clusterId),
			SgId:              string(sgId),
			SubnetId:          privateSubnets[0],
			TaskDefinitionArn: taskDefArn,
			TaskFamily:        taskDefFamily,
		})

		if err != nil {
			log.Fatal(err)
		}

		return ""
	}).(pulumi.StringOutput)

	if err != nil {
		return nil, err
	}

	return &resource, nil
}

func newContainerDefinitions(ctx *pulumi.Context, name string, args *MigrationsContainerServiceArgs, image string, options ...pulumi.ResourceOption) (pulumi.StringOutput, error) {
	logGroup, err := cloudwatch.NewLogGroup(ctx, fmt.Sprintf("%s-log-group", name), &cloudwatch.LogGroupArgs{
		NamePrefix:      pulumi.String(fmt.Sprintf("%s-pulumi-migration-logs", name)),
		RetentionInDays: pulumi.Int(1),
	}, options...)

	if err != nil {
		return pulumi.StringOutput{}, err
	}

	secrets, err := NewSecrets(ctx, fmt.Sprintf("%s-secrets", name), &SecretsArgs{
		Prefix:   args.SecretsManagerPrefix,
		KmsKeyId: args.KmsServiceKeyId,
		Secrets: []Secret{
			{
				Name:  "MYSQL_ROOT_USERNAME",
				Value: args.DatabaseArgs.Username,
			},
			{
				Name:  "MYSQL_ROOT_PASSWORD",
				Value: args.DatabaseArgs.Password,
			},
		},
	}, options...)

	if err != nil {
		return pulumi.StringOutput{}, err
	}

	taskDef, _ := pulumi.All(
		args.DatabaseArgs.ClusterEndpoint,
		args.DatabaseArgs.Port,
		secrets.Secrets,
		logGroup.ID()).ApplyT(func(applyArgs []any) (string, error) {

		dbClusterEndpoint := applyArgs[0].(string)
		dbPort := applyArgs[1].(int)
		secretsOutput := applyArgs[2].([]map[string]any)
		logId := applyArgs[3].(pulumi.ID)

		containerJson, err := json.Marshal([]any{
			map[string]any{
				"name":              "pulumi-migration",
				"image":             image,
				"cpu":               cpu,
				"memoryReservation": memoryReservation,
				"environment": []map[string]any{
					CreateEnvVar("SKIP_CREATE_DB_USER", "true"),
					CreateEnvVar("PULUMI_DATABASE_ENDPOINT", fmt.Sprintf("%s:%d", dbClusterEndpoint, dbPort)),
					CreateEnvVar("PULUMI_DATABASE_PING_ENDPOINT", dbClusterEndpoint),
				},
				"secrets": secretsOutput,
				"logConfiguration": map[string]any{
					"logDriver": "awslogs",
					"options": map[string]any{
						"awslogs-region":        args.Region,
						"awslogs-group":         logId,
						"awslogs-stream-prefix": "pulumi-api",
					},
				},
			},
		})

		if err != nil {
			return "", nil
		}

		return string(containerJson), nil
	}).(pulumi.StringOutput)

	return taskDef, nil
}

type MigrationsContainerServiceArgs struct {
	ContainerBaseArgs

	DatabaseArgs             *config.DatabaseArgs
	EcrRepoAccountId         string
	ExecuteMigrations        bool
	ImageTag                 string
	ImagePrefix              string
	SecurityGroupEgressRules ec2.SecurityGroupEgressArray
}

type MigrationsContainerService struct {
	pulumi.ResourceState

	SecurityGroup *ec2.SecurityGroup
}

File: ecs-hosted/ts/application/services/migrationsService.ts
import * as pulumi from "@pulumi/pulumi";
import * as iam from "@pulumi/aws/iam";
import * as cloudwatch from "@pulumi/aws/cloudwatch";
import * as ecs from "@pulumi/aws/ecs";
import * as ec2 from "@pulumi/aws/ec2";

import { DatabaseMigrationTask } from "../databaseMigrationTask";
import { buildECRImageTag, generateSecretsManagerPolicy } from "../utils";
import { Secrets } from "./secrets";
import { MigrationsArgs } from "./types";
import { getIamPolicyArn } from "../../common/utils";

const namespace = "pulumi:dbMigrations";
const launchType = "FARGATE";
const networkMode = "awsvpc";
const cpu = 256;
const memoryReservation = 512;

export class MigrationService extends pulumi.ComponentResource {

    public readonly securityGroup: ec2.SecurityGroup;

    private readonly baseOptions: pulumi.ComponentResourceOptions;
    private readonly baseName: string;
    private readonly baseArgs: MigrationsArgs;

    constructor(name: string, args: MigrationsArgs, opts?: pulumi.ComponentResourceOptions) {
        super(namespace, name, args, opts);

        this.baseArgs = args;
        this.baseOptions = pulumi.mergeOptions(opts, { parent: this });
        this.baseName = name;

        const image = `pulumi/migrations:${args.migrationsImageTag}`;

        const role = new iam.Role(`${name}-role`, {
            assumeRolePolicy: iam.assumeRolePolicyForPrincipal({ Service: "ecs-tasks.amazonaws.com" })
        }, this.baseOptions);

        const ecsPolicyArn = getIamPolicyArn(this.baseArgs.region, iam.ManagedPolicy.AmazonECSTaskExecutionRolePolicy);
        pulumi.log.debug(`construct ecs task execution role policy arn: ${ecsPolicyArn}`);

        new iam.RolePolicyAttachment(`${name}-task-role-attachment`, {
            role: role,
            policyArn: ecsPolicyArn
        }, this.baseOptions);

        const doc = generateSecretsManagerPolicy(args.region, args.secretsManagerPrefix!, args.kmsServiceKeyId!, args.accountId);
        new iam.RolePolicy(`${this.baseName}-migration-secrets-pol`, {
            role: role,
            policy: doc
        }, this.baseOptions);

        this.securityGroup = new ec2.SecurityGroup(`${name}-sg`, {
            vpcId: args.vpcId,
            egress: [{
                fromPort: 0,
                toPort: 0,
                protocol: "-1",
                cidrBlocks: ["0.0.0.0/0"]
            }],
        }, this.baseOptions);

        const cluster = new ecs.Cluster(`${name}-cluster`, {}, this.baseOptions);

        const taskDefinition = new ecs.TaskDefinition(`${name}-task-def`, {
            family: "pulumi-migration-task",
            networkMode: networkMode,
            cpu: cpu.toString(),
            memory: memoryReservation.toString(),
            requiresCompatibilities: [launchType],
            executionRoleArn: role.arn,
            containerDefinitions: this.constructContainerDefinitions(args, image)
        }, this.baseOptions);

        if (pulumi.runtime.isDryRun()) {
            console.log("Skipping database migration task on Pulumi Preview");

            return;
        }

        const taskMigration = new DatabaseMigrationTask(args.region);

        // create and trigger an ECS fargate task to run the db migrations
        // we are not using an ECS service as the service will attempt to restart exited containers. We only want one execution per deployment.
        pulumi
            .all([
                cluster.id,
                this.securityGroup.id,
                args.privateSubnetIds[0],
                taskDefinition.arn,
                taskDefinition.family
            ])
            .apply(async ([clusterId, securityGroupId, subnetId, taskArn, taskFamily]) => {
                await taskMigration.runMigrationTask(clusterId, securityGroupId, subnetId, taskArn, taskFamily);
            });
    }

    constructContainerDefinitions(args: MigrationsArgs, image: string): pulumi.Output<string> {
        const { database } = args;

        const logGroup = new cloudwatch.LogGroup(`${this.baseName}-log-group`, {
            name: "pulumi-migration-logs",
            retentionInDays: 1
        }, this.baseOptions);

        const migrationSecrets = new Secrets("migration-secrets", {
            prefix: args.secretsManagerPrefix,
            kmsKeyId: args.kmsServiceKeyId,
            secrets: [
                {
                    name: "MYSQL_ROOT_USERNAME",
                    value: pulumi.secret(this.baseArgs.database.dbUsername)
                },
                {
                    name: "MYSQL_ROOT_PASSWORD",
                    value: pulumi.secret(this.baseArgs.database.dbPassword)
                }
            ]
        }, this.baseOptions);

        const definition = pulumi
            .all([
                args.accountId, 
                args.ecrRepoAccountId,
                database,
                migrationSecrets.outputs,
                logGroup.id])
            .apply(([
                accountId, 
                ecrRepoAccountId,
                database, 
                secrets, 
                logId]) => {
                const ecrAccountId = ecrRepoAccountId && ecrRepoAccountId !== "" ? ecrRepoAccountId : accountId;

                // TODO: follow log groups implementation
                return JSON.stringify([{
                    name: "pulumi-migration",
                    image: buildECRImageTag(ecrAccountId, args.region, image),
                    cpu: cpu,
                    memoryReservation: memoryReservation,
                    logConfiguration: {
                        logDriver: "awslogs",
                        // http://docs.aws.amazon.com/AmazonECS/latest/developerguide/using_awslogs.html#w2ab1c21c21c13
                        options: {
                            "awslogs-region": args.region,
                            "awslogs-group": logId,
                            "awslogs-stream-prefix": "awslogs-pulumi-migration",
                        }
                    },
                    environment: [
                        {
                            name: "SKIP_CREATE_DB_USER",
                            value: "true"
                        },
                        {
                            name: "PULUMI_DATABASE_ENDPOINT",
                            value: `${database.dbClusterEndpoint}:${database.dbPort}`
                        },
                        {
                            name: "PULUMI_DATABASE_PING_ENDPOINT",
                            value: database.dbClusterEndpoint
                        }
                    ],
                    secrets: secrets
                }]);
            });

        return definition;
    }
}